### PR TITLE
refact: refactor the “prepareStringable”/”prepareRepeatable” approach

### PR DESCRIPTION
**Pull request by stone-lyl**
refact: refactor the “prepareStringable”/”prepareRepeatable” approach

image

fix: fixed the issue where param.value could not be assigned when the param.value is empty

Before

https://github.com/stone-lyl/data-story/assets/20497176/1d88ff57-356e-4935-9d8d-c55e93e113e8 

After 

https://github.com/stone-lyl/data-story/assets/20497176/a09964a5-fbf6-4a28-8bc6-d31ff5adc2af

### CODE REVIEW

File: packages/core/src/ItemWithParams/ParamsValueEvaluator.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+export class ParamEvaluator implements ParamsValueEvaluator<any>{
+  private evaluators: ParamsValueEvaluator<Param>[] = [
+    new StringableParamEvaluator(),
+    new RepeatableParamEvaluator(this),
+  ]
+
+  canEvaluate(param: Param): param is Param {
+    return this.evaluators.some(e => e.canEvaluate(param));
+  }
```
Comment (ajthinking) [inline packages/core/src/ItemWithParams/ParamsValueEvaluator.ts:15]: Linebreak 💅

File: packages/core/src/ItemWithParams/prepareRepeatable.ts [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
 import { ItemValue } from '../types/ItemValue';
-import { evalMath } from '../utils/evalMath';
-import { get } from '../utils/get';
-import Hjson from '@data-story/hjson';
-import { prepareStringable } from './prepareStringable';
+import { ParamsValueEvaluator } from '../types/ItemWithParams';
+
+export class RepeatableParamEvaluator implements ParamsValueEvaluator<RepeatableParam<any>> {
+  constructor(private evaluator: ParamsValueEvaluator<any>) {
+  }
```
Comment (ajthinking) [inline packages/core/src/ItemWithParams/prepareRepeatable.ts:7]: Prefer order: properties, constructor, public methods, private methods.

File: packages/core/src/ItemWithParams/prepareRepeatable.ts [hunk 3]
```diff
@@ -1,9 +1,17 @@
...
+  evaluate(itemValue: ItemValue, param: RepeatableParam<any>) {
+    return Object.fromEntries(param.row.map((row: {name: string}) =>
+      [row.name, this.evaluator.evaluate(itemValue, row)] as const));
+  }
+  type = 'RepeatableParam' as const;
+  canEvaluate = (param: Param): param is RepeatableParam<any> => param.type === this.type;
```
Comment (ajthinking) [inline packages/core/src/ItemWithParams/prepareRepeatable.ts:13]: In classes, prefer syntax canEvaluate(...) { ...}

**Review by ajthinking — APPROVED** [review]
Good job 👍👍 added some nitpicks 😸

File: packages/core/src/ItemWithParams/ParamsValueEvaluator.ts (hunk 1 above)
Comment (stone-lyl) [reply packages/core/src/ItemWithParams/ParamsValueEvaluator.ts:15]: suggestions really valuable 👍. I've made adjustments to the code accordingly. However, regarding the linebreak, I've made changes as shown in the image below. I hope I've understood your point correctly.

image

**Pull request merged**